skimage/util/noise.py

```diff
@@ -166,23 +166,19 @@ def random_noise(image, mode='gaussian', seed=None, clip=True, **kwargs):
                            amount=kwargs['amount'], salt_vs_pepper=0.)
 
     elif mode == 's&p':
-        # This mode makes no effort to avoid repeat sampling. Thus, the
-        # exact number of replaced pixels is only approximate.
         out = image.copy()
 
         # Salt mode
-        num_salt = np.ceil(
-            kwargs['amount'] * image.size * kwargs['salt_vs_pepper'])
-        coords = [np.random.randint(0, i - 1, int(num_salt))
-                  for i in image.shape]
-        out[coords] = 1
+        p_salt = kwargs['amount'] * kwargs['salt_vs_pepper']
+        mask = np.random.choice([True, False], size=image.shape,
+                                p=[p_salt, 1 - p_salt])
+        out[mask] = 1
 
         # Pepper mode
-        num_pepper = np.ceil(
-            kwargs['amount'] * image.size * (1. - kwargs['salt_vs_pepper']))
-        coords = [np.random.randint(0, i - 1, int(num_pepper))
-                  for i in image.shape]
-        out[coords] = low_clip
+        p_pepper = kwargs['amount'] * (1 - kwargs['salt_vs_pepper'])
+        mask = np.random.choice([True, False], size=image.shape,
+                                p=[p_pepper, 1 - p_pepper])
+        out[mask] = low_clip
```

Comment:
@soupault it never did guarantee that, nor would you want to. It's probabilistic. What you want is the probability of a pixel being flipped equal to `kwargs['amount']`. Having said that, currently that's not true. I think the right way to do this is to generate a "flipped" mask and then a "salt" mask and "and" them together, then "and" "flipped" with "not salt". Working on it.